Update command injection docs (#741)
returntocorp/semgrep-docs: docs/cheat-sheets/java-command-injection.mdx

---
slug: java-command-injection
---

import LinkToRegistryRule from "/src/components/LinkToRegistryRule"
import CommandInjectionIntro from "/src/components/concept/_command-injection-intro.mdx"

# Command injection prevention for Java

<CommandInjectionIntro />

#### Check your project for these conditions:

```bash
semgrep scan --config p/java-command-injection
```

---

## 1. Running an OS command

### 1.A. Running OS commands with Runtime.getRuntime().exec()

> The `exec` executes the specified string command in a separate process. This is dangerous if a command string is controlled by user input and could result in a command injection.

Reviewer: the `exec` call?

Example:

```java
 // exec example
Runtime.getRuntime().exec("ls -la")

// Vunerable
String input = "&& cat /etc/passwd"; // value supplied by user input
Runtime r = Runtime.getRuntime();
r.exec("some_tool -t param1 param2 " + input);

// Vunerable
String input = "cat /etc/passwd"; // value supplied by user input
Runtime.getRuntime().exec("bash", "-c", input);
```

#### References

- [`exec` documentation](https://docs.oracle.com/javase/7/docs/api/java/lang/Runtime.html#exec(java.lang.String))

#### Mitigation

- Always try to use internal Java API (if it exists) instead of running an OS command. In other words, use internal language features instead of invoking commands that can be exploited.
- Do not include command arguments in a command string, use parameterization instead. For example:<br />
    Use:
    ```java
    Runtime.getRuntime().exec("/path/to/myCommand", "myArg1", inputValue)
    ```
    Instead of:
    ```java
    Runtime.getRuntime().exec("bash", "-c", "myCommand myArg1 " + inputValue)
    ```
- If it is not possible, then strip the input of everything except alphanumeric characters provided for the command string and arguments.

> - Try to avoid non-literal values for the command string, such as `Runtime.getRuntime().exec("bash", "-c", input)` in the example above. Do not use `input` and similar functions.

Reviewer: doesn't seem like 'input' is a function here, so maybe something like:
Beware of using direct user input, even if it is sanitized.

Author: ```suggestion
- Do **not** use direct user input, even if it is sanitized.
```

> - If it is not possible, then do not let running arbitrary commands, use an allowlist for inputs.

Reviewer: `let arbitrary commands run`

> - Strip `!@#$;&*~"'{}][-+%^` characters from an input provided for command string. 

Reviewer: `from user input that is incorporated in the command string which is later executed.`

Author: ```suggestion
- Strip `!@#$;&*~"'{}][-+%^` characters from user input that is incorporated in the command string which is later executed.
```

#### Semgrep rule

<LinkToRegistryRule ruleId="java.lang.security.audit.command-injection-formatted-runtime-call.command-injection-formatted-runtime-call" />

### 1.B. Running OS processes with ProcessBuilder

> ProcessBuilder class is used to create operating system processes. If the command string is controlled by user input it could result in command injection. 

Author: ```suggestion
ProcessBuilder class is used to create operating system processes. If the command string is controlled by user input it can result in command injection. 
```

Example:

```java
 // ProcessBuilder example
Process builder = new ProcessBuilder("ls", "-la").start();

// Vunerable
String input = "cat /etc/passwd"; // value supplied by user input
ProcessBuilder builder = new ProcessBuilder();
builder.command("bash", "-c", "some_tool -t param1 param2 " + input)

// Vulnerable
String input = "cat /etc/passwd"; // value supplied by user input
ProcessBuilder builder = new ProcessBuilder();
builder.command(input);
```

#### References

[`ProcessBuilder` documentation](https://docs.oracle.com/javase/7/docs/api/java/lang/ProcessBuilder.html)

#### Mitigation

> - Try to avoid non-literal values for the command string.

Reviewer: 'in the command string' instead

Author: ```suggestion
- Try to avoid non-literal values in the command string.
```

> - If it is not possible, then do not let running arbitrary commands, use an allowlist for inputs.

Reviewer: `do not let arbitrary commands run`

Author: @colleend I feel that "do not let arbitrary commands run" would be a sentence fragment without subsequent details (for example: do not let arbitrary commands run in X and B). Maybe:
```suggestion
- If it is not possible to prevent non-literal values in the command string, then do not allow running arbitrary commands. Use an allowlist for inputs.
```

---

import LinkToRegistryRule from "/src/components/LinkToRegistryRule"
import CommandInjectionIntro from "/src/components/concept/_command-injection-intro.mdx"

# Command injection prevention for Java

<CommandInjectionIntro />

#### Check your project for these conditions:

```bash
semgrep --config p/java-command-injection
```

---

## 1. Running an OS command

### 1.A. Running OS commands with `Runtime.getRuntime().exec()`

`exec` executes the specified string command in a separate process. This is dangerous if a command string is controlled by user input and could result in a command injection.

Example:

```java

 // exec example
Runtime.getRuntime().exec("ls -la")

// Vunerable
String input = "&& cat /etc/passwd"; // value supplied by user input
Runtime r = Runtime.getRuntime();
r.exec("some_tool -t param1 param2 " + input);

// Vunerable
String input = "cat /etc/passwd"; // value supplied by user input
Runtime.getRuntime().exec("bash", "-c", input);
```

#### References

- [`exec` documentation](https://docs.oracle.com/javase/7/docs/api/java/lang/Runtime.html#exec(java.lang.String))

#### Mitigation

- Try to avoid non-literal values for the command string.

> - If it is not possible, then do not let running arbitrary commands, use a white list for inputs.

Author: ```suggestion
- If it is not possible, then do not let running arbitrary commands, use an allow list for inputs.
```

> - Try to avoid non-literal values for the command string.
> - If it is not possible, then do not let running arbitrary commands, use an allowlist for inputs.
> - Strip `!@#$;&*~"'{}][-+%^` characters from an input provided for command string. 

Author: ```suggestion
- Always try to use internal Java API (if it exists) instead of running an OS command. In other words, use internal language features instead of invoking commands that can be exploited.
- Avoid running a shell as a command with arguments, if it is not possible, then strip the input of everything except alphanumeric characters provided for the command string and arguments.
- Try to avoid non-literal values for the command string, such as `Runtime.getRuntime().exec("bash", "-c", input)` in the example above. Do not use `input` and similar functions.
- If it is not possible, then do not let running arbitrary commands, use an allowlist for inputs.
- Strip `!@#$;&*~"'{}][-+%^` characters from an input provided for command string. 
```

#### Semgrep rule

<LinkToRegistryRule ruleId="java.lang.security.audit.command-injection-formatted-runtime-call.command-injection-formatted-runtime-call" />

### 1.B. Running OS processes with ProcessBuilder

ProcessBuilder class is used to create operating system processes. If the command string is controlled by user input it could result in command injection. 
Example:

```java
 // ProcessBuilder example
Process builder = new ProcessBuilder("ls", "-la").start();

// Vunerable
String input = "cat /etc/passwd"; // value supplied by user input
ProcessBuilder builder = new ProcessBuilder();
builder.command("bash", "-c", "some_tool -t param1 param2 " + input)

// Vulnerable
String input = "cat /etc/passwd"; // value supplied by user input
ProcessBuilder builder = new ProcessBuilder();
builder.command(input);
```

#### References

[`ProcessBuilder` documentation](https://docs.oracle.com/javase/7/docs/api/java/lang/ProcessBuilder.html)

#### Mitigation

- Try to avoid non-literal values for the command string.

> - If it is not possible, then do not let running arbitrary commands, use a white list for inputs.

Author: ```suggestion
- If it is not possible, then do not let running arbitrary commands, use an allow list for inputs.
```
